### PR TITLE
Implement BigInt spec support

### PR DESCRIPTION
**Pull request by frostney**
## Summary
- Added native BigInt support with parsing, arithmetic, bitwise operations, comparisons, shifts, conversions, and `asIntN` / `asUintN` helpers.
- Wired BigInt into the runtime, lexer, parser, compiler, VM, JSON handling, error messages, and global built-ins.
- Added BigInt literals, coercion, constructor, prototype-method, arithmetic, bitwise, comparison, and type-error coverage.

### CODE REVIEW

File: source/shared/BigInteger.pas [hunk 1]
```diff
@@ -0,0 +1,1206 @@
...
+    function Divide(const AOther: TBigInteger): TBigInteger;
+    function Modulo(const AOther: TBigInteger): TBigInteger;
+    function Power(const AExponent: TBigInteger): TBigInteger;
+
+    function BitwiseAnd(const AOther: TBigInteger): TBigInteger;
+    function BitwiseOr(const AOther: TBigInteger): TBigInteger;
+    function BitwiseXor(const AOther: TBigInteger): TBigInteger;
+    function BitwiseNot: TBigInteger;
+    function ShiftLeft(const ACount: Integer): TBigInteger;
+    function ShiftRight(const ACount: Integer): TBigInteger;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/shared/BigInteger.pas:60]: The current `Int64` parameter for shift counts is intentional. Shift counts beyond 2^63-1 bits are impractical (that would require exabytes of memory for the result). The `Int64` range covers all realistic use cases and avoids the overhead of arbitrary-precision shift count parsing.

File: source/units/Goccia.Builtins.GlobalBigInt.pas [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  if Arg is TGocciaNumberLiteralValue then
+  begin
+    NumVal := TGocciaNumberLiteralValue(Arg).Value;
+    if TGocciaNumberLiteralValue(Arg).IsNaN or
+       TGocciaNumberLiteralValue(Arg).IsInfinite then
+      ThrowTypeError(Format(SErrorBigIntNotInteger, [Arg.ToStringLiteral.Value]),
+        SSuggestBigIntNoImplicitConversion);
+    if Frac(NumVal) <> 0 then
+      ThrowTypeError(Format(SErrorBigIntNotInteger, [Arg.ToStringLiteral.Value]),
+        SSuggestBigIntNoImplicitConversion);
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Builtins.GlobalBigInt.pas:126]: Fixed in 000e9636 — changed to `ThrowRangeError` per ES2026 §21.2.1.1.1 NumberToBigInt.

File: source/units/Goccia.Builtins.GlobalBigInt.pas [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  begin
+    try
+      Result := TGocciaBigIntValue.Create(
+        TBigInteger.FromDecimalString(Trim(TGocciaStringLiteralValue(Arg).Value)));
+    except
+      ThrowTypeError(Format(SErrorBigIntInvalidConversion, [
+        '''' + TGocciaStringLiteralValue(Arg).Value + '''']),
+        SSuggestBigIntNoImplicitConversion);
+      Result := nil;
+    end;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Builtins.GlobalBigInt.pas]: Fixed in 000e9636 — `BigInt()` constructor now routes `0x`/`0o`/`0b` prefixed strings to the corresponding `FromHexString`/`FromOctalString`/`FromBinaryString` parsers, and throws `SyntaxError` (not `TypeError`) for invalid string inputs.

File: source/units/Goccia.Builtins.GlobalBigInt.pas [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+  Bits := Trunc(AArgs.GetElement(0).ToNumberLiteral.Value);
+
+  BigIntArg := AArgs.GetElement(1);
+  if not (BigIntArg is TGocciaBigIntValue) then
+    ThrowTypeError(Format(SErrorBigIntInvalidConversion, [BigIntArg.TypeName]),
+      SSuggestBigIntNoImplicitConversion);
+
+  Result := TGocciaBigIntValue.Create(
+    TGocciaBigIntValue(BigIntArg).Value.AsIntN(Bits));
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Builtins.GlobalBigInt.pas:223]: `asIntN`/`asUintN` are explicitly listed as not implemented and deferred. The current stub code is placeholder only — `ToIndex` coercion will be addressed when these methods are fully implemented.

File: source/units/Goccia.Engine.pas [hunk 5]
```diff
@@ -695,6 +701,7 @@ procedure TGocciaEngine.RegisterBuiltIns;
   FBuiltinGlobalObject := TGocciaGlobalObject.Create(CONSTRUCTOR_OBJECT, Scope, ThrowError);
   FBuiltinGlobalArray := TGocciaGlobalArray.Create(CONSTRUCTOR_ARRAY, Scope, ThrowError);
   FBuiltinGlobalNumber := TGocciaGlobalNumber.Create(CONSTRUCTOR_NUMBER, Scope, ThrowError);
+  FBuiltinGlobalBigInt := TGocciaGlobalBigInt.Create(CONSTRUCTOR_BIGINT, Scope, ThrowError);
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Engine.pas:705]: Fixed in 000e9636 — added `FBuiltinGlobalBigInt.Free` to `TGocciaEngine.Destroy`.

File: source/units/Goccia.Evaluator.Arithmetic.pas [hunk 6]
```diff
@@ -36,11 +39,19 @@ function InfinityWithSign(const APositive: Boolean): TGocciaNumberLiteralValue;
...
+    ThrowTypeError(SErrorBigIntMixedTypes, SSuggestBigIntNoMixedArithmetic);
+end;
+
 function ToNumericPair(const ALeft, ARight: TGocciaValue;
   out ALeftNum, ARightNum: TGocciaNumberLiteralValue): Boolean;
 begin
   if (ALeft is TGocciaSymbolValue) or (ARight is TGocciaSymbolValue) then
     ThrowTypeError(SErrorSymbolToNumber, SSuggestSymbolNoImplicitConversion);
+  CheckBigIntMixedTypes(ALeft, ARight);
   ALeftNum := ALeft.ToNumberLiteral;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Evaluator.Arithmetic.pas:55]: GocciaScript does not support BigInt object wrappers (`Object(1n)`). The `Box` method in `ClassHelper.pas` returns `nil` for BigInt primitives, and there is no `BigIntObjectValue` type. `ToPrimitive` is therefore a no-op for BigInt in practice. If BigInt boxing is added in the future, these paths will need updating.

File: source/units/Goccia.Evaluator.Arithmetic.pas [hunk 7]
```diff
@@ -158,6 +196,16 @@ function EvaluateDivision(const ALeft, ARight: TGocciaValue): TGocciaValue;
...
   SameSign: Boolean;
 begin
+  // ES2026 §6.1.6.2.6 BigInt::divide — truncates toward zero
+  if (ALeft is TGocciaBigIntValue) and (ARight is TGocciaBigIntValue) then
+  begin
+    if TGocciaBigIntValue(ARight).Value.IsZero then
+      ThrowTypeError(SErrorBigIntDivisionByZero, SSuggestBigIntNoMixedArithmetic);
+    Result := TGocciaBigIntValue.Create(
+      TGocciaBigIntValue(ALeft).Value.Divide(TGocciaBigIntValue(ARight).Value));
+    Exit;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Evaluator.Arithmetic.pas:218]: Fixed in 000e9636 — changed BigInt division-by-zero, modulo-by-zero, and negative exponent errors to `ThrowRangeError` in both the evaluator and VM.

File: source/units/Goccia.Evaluator.Comparison.pas [hunk 8]
```diff
@@ -520,7 +537,38 @@ function CompareNumbers(const ALeftNum, ARightNum: TGocciaNumberLiteralValue; co
...
+  end;
+  // Compare via Double (may lose precision for very large BigInts, but matches spec behavior)
+  D := ABigInt.Value.ToDouble;
+  if D > ANumber.Value then
+    Result := 1
+  else if D < ANumber.Value then
+    Result := -1
+  else
+    Result := 0;
+end;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Evaluator.Comparison.pas]: Acknowledged — `ToDouble` comparison loses precision beyond 2^53. This is a known limitation. An exact mathematical-value comparison (comparing BigInt magnitude against the Number's integer part limb-by-limb) would fix it but adds significant complexity. The current approach matches the practical behavior of most lightweight runtimes and covers the common case. We can revisit if test262 Temporal tests surface precision failures.

File: source/units/Goccia.Evaluator.Comparison.pas [hunk 9]
```diff
@@ -562,16 +631,22 @@ function GreaterThanOrEqual(const ALeft, ARight: TGocciaValue): Boolean;
...
+  if not ((ALeft is TGocciaBigIntValue) or (ARight is TGocciaBigIntValue)) then
   begin
-    Result := False;
-    Exit;
+    if ALeft.ToNumberLiteral.IsNaN or ARight.ToNumberLiteral.IsNaN then
+    begin
+      Result := False;
+      Exit;
+    end;
   end;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Evaluator.Comparison.pas]: Fixed in 000e9636 — `GreaterThanOrEqual` and `LessThanOrEqual` now explicitly check the Number operand for NaN in cross-type BigInt/Number comparisons, in both the evaluator and VM. `1n >= NaN` and `1n <= NaN` now correctly return `false`.

File: source/units/Goccia.JSON.pas [hunk 10]
```diff
@@ -837,6 +838,10 @@ function TGocciaJSONStringifier.StringifyPreparedValue(const AValue: TGocciaValu
     else
       Result := '"' + EscapeJSONString(EffectiveValue.ToStringLiteral.Value) + '"';
   end
+  // ES2026 §25.5.2.5 step 10: BigInt values throw TypeError
+  else if EffectiveValue is TGocciaBigIntValue then
+    ThrowTypeError('Do not know how to serialize a BigInt',
+      'use BigInt.prototype.toString() or a custom replacer')
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.JSON.pas:844]: The `toJSON` hook for BigInt primitives is an edge case that requires infrastructure changes to `ApplyToJSON` (currently only probes objects). Since `JSON.stringify(1n)` throwing `TypeError` is the standard behavior when no `toJSON` is defined, and defining `BigInt.prototype.toJSON` is a userland pattern, this can be addressed in a follow-up if needed.

File: source/units/Goccia.Lexer.pas [hunk 11]
```diff
@@ -1421,10 +1426,22 @@ procedure TGocciaLexer.ScanNumber;
...
-  AddToken(gttNumber, Lexeme);
+  // ES2026 §12.9.3 BigIntLiteralSuffix: check for 'n' suffix
+  if (Peek = 'n') and (not HasDecimalOrExponent) then
+  begin
+    Advance;
+    Lexeme := Copy(FSource, FStart, FCurrent - FStart - 1); // Exclude 'n'
+    if HasSeparator then
+      Lexeme := StringReplace(Lexeme, '_', '', [rfReplaceAll]);
+    AddToken(gttBigInt, Lexeme);
+  end
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Lexer.pas:1437]: This is a false positive. The parser does **not** use `FromDecimalString` for BigInt literals. `ConvertBigIntLiteral` (Parser.pas line 4339) already dispatches to `FromHexString`, `FromBinaryString`, and `FromOctalString` based on the prefix. The lexer/parser contract is correct.

File: source/units/Goccia.Values.BigIntValue.pas [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+  Radix := 10;
+  if AArgs.Length > 0 then
+  begin
+    RadixValue := AArgs.GetElement(0);
+    if not (RadixValue is TGocciaUndefinedLiteralValue) then
+    begin
+      Radix := Trunc(RadixValue.ToNumberLiteral.Value);
+      if (Radix < 2) or (Radix > 36) then
+        ThrowTypeError('toString() radix must be between 2 and 36',
+          'provide a radix value between 2 and 36');
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Values.BigIntValue.pas]: Fixed in 000e9636 — changed to `ThrowRangeError` for invalid `toString()` radix values.

File: source/units/Goccia.VM.pas [hunk 13]
```diff
@@ -1754,6 +1799,9 @@ function TGocciaVM.ConstantToValue(const AConstant: TGocciaBytecodeConstant): TG
       Result := TGocciaNumberLiteralValue.Create(AConstant.FloatValue);
     bckString:
       Result := TGocciaStringLiteralValue.Create(AConstant.StringValue);
+    bckBigInt:
+      Result := TGocciaBigIntValue.Create(
+        TBigInteger.FromDecimalString(AConstant.StringValue));
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.VM.pas:1955]: Fixed in 000e9636 — all VM inline helpers (`VMBitwiseAndValues`, `VMBitwiseOrValues`, `VMBitwiseXorValues`, `VMLeftShiftValues`, `VMRightShiftValues`, `VMUnsignedRightShiftValues`, `VMBitwiseNotValue`, `VMLessThan`, `VMGreaterThan`, `VMLessThanOrEqual`, `VMGreaterThanOrEqual`) and `OP_NEG` now have BigInt dispatch. `IsEqual` in `ClassHelper` also handles BigInt strict equality. All 51 BigInt tests pass in bytecode mode.

File: tests/language/expressions/bigint/bigint-comparison.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
+/*---
+description: BigInt comparison operations
+features: [bigint]
+---*/
+
+test("strict equality between BigInts", () => {
+  expect(1n === 1n).toBe(true);
+  expect(1n === 2n).toBe(false);
+  expect(0n === 0n).toBe(true);
```
[the remark this answers is not in the document]
Comment (frostney) [reply tests/language/expressions/bigint/bigint-comparison.js]: Fixed in 000e9636 — replaced literal self-comparisons with distinct variable bindings using `BigInt()` constructor to preserve the test semantics.

File: source/units/Goccia.Builtins.GlobalBigInt.pas (hunk 4 above)
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Builtins.GlobalBigInt.pas:223]: Fixed in 477b5609 — `asIntN`/`asUintN` now use proper `ToIndex` coercion for the `bits` parameter: `undefined` → 0, `NaN` → 0, `Infinity` → `RangeError`, negative → `RangeError`. Tests added for all edge cases.

File: source/units/Goccia.Evaluator.Arithmetic.pas (hunk 6 above)
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Evaluator.Arithmetic.pas:55]: Verified that primitive BigInt method dispatch already works without boxing — `(1n).toString()` correctly returns `"1"` in both interpreter and bytecode modes. `TGocciaBigIntValue.GetProperty` delegates to the shared prototype, so `ToPrimitive` is not needed for method calls on BigInt primitives. The `Object(1n)` wrapping pattern is not supported by GocciaScript (no `BigIntObjectValue` type exists).

File: source/units/Goccia.Evaluator.Comparison.pas (hunk 8 above)
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Evaluator.Comparison.pas]: Fixed in 477b5609 — replaced the `ToDouble`-based comparison with exact mathematical-value comparison. Integer Numbers are converted to `TBigInteger` via `FromDouble` and compared limb-by-limb. Fractional Numbers are compared against their floor/ceil. `9007199254740993n > 9007199254740992` now correctly returns `true`. Test added for the 2^53 edge case.

File: source/units/Goccia.Evaluator.Arithmetic.pas [hunk 15]
```diff
@@ -250,6 +308,16 @@ function EvaluateExponentiation(const ALeft, ARight: TGocciaValue): TGocciaValue
 var
   LeftNum, RightNum: TGocciaNumberLiteralValue;
 begin
+  // ES2026 §6.1.6.2.8 BigInt::exponentiate
+  if (ALeft is TGocciaBigIntValue) and (ARight is TGocciaBigIntValue) then
+  begin
+    if TGocciaBigIntValue(ARight).Value.IsNegative then
+      ThrowRangeError(SErrorBigIntNegativeExponent);
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Evaluator.Arithmetic.pas:335]: Fixed in d59f8ea9 — changed to "Exponent must be non-negative".

File: source/units/Goccia.Values.ClassHelper.pas [hunk 16]
```diff
@@ -33,7 +33,11 @@   TGocciaValueHelper = class helper for TGocciaValue
 implementation
 
 uses
-  Math;
+  Math,
+
+  BigInteger,
+
+  Goccia.Values.BigIntValue;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Values.ClassHelper.pas:41]: BigInt strict equality is covered by 8 JS-level test files (61 tests, 116 assertions) including dedicated comparison tests for value equality, cross-type inequality, NaN, fractional numbers, and precision beyond 2^53. The `IsEqual` class helper change mirrors the existing pattern for Number/String/Boolean which also lack dedicated Pascal unit tests — the JS test suite serves as the integration test layer for the value system.

File: source/units/Goccia.Evaluator.Arithmetic.pas (hunk 6 above)
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Evaluator.Arithmetic.pas:55]: Fixed in 3344dede — implemented `TGocciaBigIntObjectValue` for BigInt boxing via `Object(1n)`. Added `ToPrimitive` calls to subtraction, multiplication, division, modulo, and exponentiation so `Object(1n) - 1n === 0n`, `Object(2n) * 3n === 6n`, and `Object(2n) ** 3n === 8n` all work correctly. Also fixed `Object()` to call `Box` for primitive arguments per ES2026 §20.1.1.1.

File: source/units/Goccia.Values.BigIntObjectValue.pas [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+function TGocciaBigIntObjectValue.BigIntObjectValueOf(const AArgs: TGocciaArgumentsCollection; const AThisValue: TGocciaValue): TGocciaValue;
+begin
+  if AThisValue is TGocciaBigIntObjectValue then
+    Result := TGocciaBigIntObjectValue(AThisValue).Primitive
+  else if AThisValue is TGocciaBigIntValue then
+    Result := AThisValue
+  else
+    Result := AThisValue;
+end;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Values.BigIntObjectValue.pas:122]: Fixed in a9c40268 — `valueOf` now throws `TypeError` for non-BigInt/non-BigInt-object receivers instead of returning the arbitrary `this`. `toString` validates the receiver, parses the optional radix from `AArgs[0]` (defaulting to 10), validates it is in `[2, 36]` (throws `RangeError` via `SErrorBigIntInvalidRadix`), and produces radix-based output via `ToRadixString`. Uses the same centralized error constants as the primitive prototype methods.
